### PR TITLE
i#2399 add reuse time tool to drmemtrace.

### LINKED ISSUE #2399: add reuse time tool to drmemtrace
Xref #2300 

This issue covers creating a "reuse time" tool, which just calculates # references between consecutive accesses to the same address, not # unique addresses in between. Could either be a separate (and cheaper) tool, or just an option to reuse_distance_t.

Most importantly, there should be a programatic way to access the output, so we can use it in downstream tools outside of the DR framework. I can take a stab at the implementation, but I'm not sure what's the best interface change to analysis_tool_t to allow for this. Any ideas?

### PR DESCRIPTION
**Pull request by s-kanev**
Adds a separate reuse time tool, which grabs a histogram of # memory
accesses between accesses to the same address. Space and time use is
linear with the size of the trace.

TESTED: ctest; code_api|sample.* were failing before.

Fixes #2399

### CODE REVIEW

File: suite/tests/CMakeLists.txt [hunk 1]
```diff
@@ -2473,6 +2473,14 @@ if (CLIENT_INTERFACE)
           "-infile ${small_trace_file} -simulator_type reuse_distance -reuse_distance_histogram" "" "")
         set(tool.reuse.offline_toolname "drcachesim")
         set(tool.reuse.offline_basedir "${PROJECT_SOURCE_DIR}/clients/drcachesim/tests")
+
+        set(small_trace_file
+          "${PROJECT_SOURCE_DIR}/clients/drcachesim/tests/drmemtrace.small.x64.trace")
```
Comment (derekbruening) [inline suite/tests/CMakeLists.txt]: Redundant set, right?

File: clients/drcachesim/tools/reuse_time.cpp [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+reuse_time_t::reuse_time_t(unsigned int line_size, unsigned int verbose) :
+    time_stamp(0), knob_verbose(verbose), knob_line_size(line_size)
+{
+    line_size_bits = compute_log2((int)knob_line_size);
+}
+
+reuse_time_t::~reuse_time_t() {}
+
+bool reuse_time_t::process_memref(const memref_t &memref)
```
Comment (derekbruening) [inline clients/drcachesim/tools/reuse_time.cpp]: style: return type on own line

File: clients/drcachesim/tools/reuse_time.cpp [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    return new reuse_time_t(line_size, verbose);
+}
+
+reuse_time_t::reuse_time_t(unsigned int line_size, unsigned int verbose) :
+    time_stamp(0), knob_verbose(verbose), knob_line_size(line_size)
+{
+    line_size_bits = compute_log2((int)knob_line_size);
+}
+
+reuse_time_t::~reuse_time_t() {}
```
Comment (derekbruening) [inline clients/drcachesim/tools/reuse_time.cpp]: style: braces on own lines (with sthg like "// Empty." is the convention)

File: clients/drcachesim/tools/reuse_time.cpp [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+    // Only care about data for now.
+    if (type_is_instr(memref.instr.type)) {
+        return true;
+    }
+
+    // Ignore thread events and other tracing metadata.
+    if (memref.data.type != TRACE_TYPE_READ &&
+        memref.data.type != TRACE_TYPE_WRITE &&
+        type_is_prefetch(memref.data.type)) {
```
Comment (derekbruening) [inline clients/drcachesim/tools/reuse_time.cpp]: I think you meant ! for the prefetch?

File: suite/tests/CMakeLists.txt (hunk 1 above)
Comment (s-kanev) [inline suite/tests/CMakeLists.txt]: Done. Yup, just blind copy-paste.

File: clients/drcachesim/tools/reuse_time.cpp (hunk 3 above)
Comment (s-kanev) [inline clients/drcachesim/tools/reuse_time.cpp]: Done.

File: clients/drcachesim/tools/reuse_time.cpp (hunk 2 above)
Comment (s-kanev) [inline clients/drcachesim/tools/reuse_time.cpp]: Done.

File: clients/drcachesim/tools/reuse_time.cpp (hunk 4 above)
Comment (s-kanev) [inline clients/drcachesim/tools/reuse_time.cpp]: Done. Thanks for catching!

**Pull request merged**